### PR TITLE
Fix `torch.compile` with `fullgraph=True` when `attention_mask` input is used

### LINKED ISSUE #29190: torch.export fails for llama model
### System Info

- `transformers` version: 4.38.0.dev0
- Platform: Linux-5.12.0-0_fbk15_zion_7536_ged335dd7211d-x86_64-with-glibc2.34
- Python version: 3.11.7
- Huggingface_hub version: 0.20.3
- Safetensors version: 0.4.2
- Accelerate version: 0.27.2
- Accelerate config:    not found
- PyTorch version (GPU?): 2.3.0.dev20240221+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

@ArthurZucker @gante

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

llama = AutoModelForCausalLM.from_pretrained(
        "meta-llama/Llama-2-7b-chat-hf", low_cpu_mem_usage=True
    )
tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-chat-hf")

llama.eval()
inputs = tokenizer("How do you", return_tensors="pt")

exported_llama = torch.export.export(llama, args=(inputs["input_ids"],inputs["attention_mask"]))
```

### Expected behavior

Successfully export of the model but fails with:
```
torch._dynamo.exc.UserError: Dynamic control flow is not supported at the moment. Please use functorch.experimental.control_flow.cond to explicitly capture the control flow. For more information about this error, see: https://pytorch.org/docs/main/generated/exportdb/index.html#cond-operands

from user code:
   File "/home/mreso/.conda/envs/pippy/lib/python3.11/site-packages/transformers/models/llama/modeling_llama.py", line 1176, in forward
    outputs = self.model(
  File "/home/mreso/.conda/envs/pippy/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1538, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/mreso/.conda/envs/pippy/lib/python3.11/site-packages/transformers/models/llama/modeling_llama.py", line 990, in forward
    causal_mask = self._update_causal_mask(attention_mask, inputs_embeds)
  File "/home/mreso/.conda/envs/pippy/lib/python3.11/site-packages/transformers/models/llama/modeling_llama.py", line 1090, in _update_causal_mask
    if not is_tracing and attention_mask is not None and torch.any(attention_mask != 1):
```
I worked around this issue by guarding with torch._dynamo_is_compiling() in these two places:
https://github.com/huggingface/transformers/blob/fc37f38915372c15992b540dfcbbe00a916d4fc6/src/transformers/models/llama/modeling_llama.py#L1080
```
is_tracing = torch.jit.is_tracing() or isinstance(input_tensor, torch.fx.Proxy) or (hasattr(torch,"_dynamo") and not torch._dynamo.is_compiling())
```
https://github.com/huggingface/transformers/blob/fc37f38915372c15992b540dfcbbe00a916d4fc6/src/transformers/models/llama/modeling_llama.py#L131
```
if hasattr(torch,"_dynamo") and not torch._dynamo.is_compiling():
    self._cos_cached = cos
    self._sin_cached = sin
```
Happy to create a PR if this is a viable solution.

### PR DESCRIPTION
**Pull request by fxmarty**
As per title.

Fixes https://github.com/huggingface/transformers/issues/29190

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -1076,10 +1076,18 @@ def _update_causal_mask(self, attention_mask, input_tensor):
...
-            is_tracing = torch.jit.is_tracing() or isinstance(input_tensor, torch.fx.Proxy)
-            if not is_tracing and attention_mask is not None and torch.any(attention_mask != 1):
-                causal_mask = causal_mask.mul(~torch.all(causal_mask == causal_mask.min(), dim=-1)[..., None]).to(
+        if self.config._attn_implementation == "sdpa" and attention_mask is not None:
+            # TODO: For dynamo, rather use a check on fullgraph=True once this is possible (https://github.com/pytorch/pytorch/pull/120400).
+            is_tracing = (
+                torch.jit.is_tracing()
+                or isinstance(input_tensor, torch.fx.Proxy)
+                or (hasattr(torch, "_dynamo") and torch._dynamo.is_compiling())
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1085]: 🤢

**Review by ArthurZucker — COMMENTED** [review]
I guess we don't have a choice?

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -1076,10 +1076,18 @@ def _update_causal_mask(self, attention_mask, input_tensor):
...
+            if not is_tracing and torch.any(attention_mask != 1):
+                # Attend to all tokens in masked rows from the causal_mask, for example the relevant first rows when
+                # using left padding. This is required by F.scaled_dot_product_attention memory-efficient attention path.
+                # Details: https://github.com/pytorch/pytorch/issues/110213
+                causal_mask = causal_mask.mul(~torch.all(causal_mask == causal_mask.min(), dim=-1, keepdim=True)).to(
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply src/transformers/models/llama/modeling_llama.py:1090]: @ArthurZucker this would conflict but is unrelated

**Comment by fxmarty** [comment]
> I guess we don't have a choice?

The other choice would be:

```python
            is_tracing = (
                torch.jit.is_tracing()
                or isinstance(input_tensor, torch.fx.Proxy)
                or torch._dynamo.is_fullgraph_tracing())
            )
```

but `torch._dynamo.is_fullgraph_tracing` does not exist in PyTorch.

One other possibility is to always do `causal_mask = causal_mask.mul(~torch.all(causal_mask == causal_mask.min(), dim=-1, keepdim=True))` no matter what.

One other possibility is to not use `causal_mask = causal_mask.mul(~torch.all(causal_mask == causal_mask.min(), dim=-1, keepdim=True))` at all, but that means that we drop support for memory-efficient attention backend in Transformers cc @drisspg

**Review by ArthurZucker — APPROVED** [review]
Alright, let's make sure CIs are green en bench are not slower!

**Comment by kwen2501** [comment]
Thanks for the fix!
Just wanted to share new API offerings from PyTorch (as of a couple days ago):
https://github.com/pytorch/pytorch/pull/119602

Summary:
- A more general flag is named `torch.compiler.is_compiling()`. This flag indicates whether a graph is traced/compiled via `torch.export()` or `torch.compile()`. The flag works even in non-strict mode (i.e. even without TorchDynamo).
- A more specific flag is named `torch.compiler.is_dynamo_compiling()`, it's stricter, because it's only set to True when TorchDynamo is used, so, in non-strict mode it would be False.

Cc: @mreso @khabinov

**Comment by fxmarty** [comment]
@kwen2501 Thank you! torch.export is not always using dynamo? Reading https://pytorch.org/docs/stable/export.html I thought so!

**Pull request merged**